crates/binjs_io/src/context/huffman.rs

```diff
@@ -0,0 +1,252 @@
+    ///
+    /// # Requirement
+    ///
+    /// Values of `T` in the source MUST be distinct.
+    pub fn from_instances<S>(source: S) -> Self
+    where
+        S: IntoIterator<Item = (T, Instances)>,
+    {
+        let mut bit_lengths = Self::compute_bit_lengths(source);
+
+        // Canonicalize order: (BitLen, T)
+        // As values of `T` are
+        bit_lengths.sort_unstable_by_key(|&(ref value, ref bit_len)| (*bit_len, value.clone()));
+
+        // The bits associated to the next value.
+        let mut bits = 0;
+        let mut keys = Vec::with_capacity(bit_lengths.len());
+
+        for i in 0..bit_lengths.len() - 1 {
+            let (bit_len, symbol, next_bit_len) = (
+                bit_lengths[i].1,
+                bit_lengths[i].0.clone(),
+                bit_lengths[i + 1].1,
+            );
+            keys.push((symbol.clone(), Key { bits, bit_len }));
+            bits = (bits + 1) << (next_bit_len - bit_len);
+        }
+        // Handle the last element.
+        let (ref symbol, bit_len) = bit_lengths[bit_lengths.len() - 1];
+        keys.push((symbol.clone(), Key { bits, bit_len }));
+
+        return Self { keys };
+    }
+
+    /// Convert a sequence of values labelled by their number of instances
+    /// into a sequence of values labelled by the length for their path
+    /// in the Huffman tree, aka the bitlength of their Huffman key.
+    ///
+    /// Values that have 0 instances are skipped.
+    pub fn compute_bit_lengths<S>(source: S) -> Vec<(T, BitLen)>
+    where
+        S: IntoIterator<Item = (T, Instances)>,
+    {
+        // Build a min-heap sorted by number of instances.
+        use std::cmp::Reverse;
+        let mut heap = BinaryHeap::new();
+
+        // Skip values that have 0 instances.
+        for (value, instances) in source {
+            if !instances.is_zero() {
+                heap.push(Reverse(Node {
+                    instances,
+                    content: NodeContent::Leaf(value),
+                }));
+            }
+        }
+
+        let len = heap.len();
+        if len == 0 {
+            // Special case: no tree to build.
+            return vec![];
+        }
+
+        // Take the two rarest nodes, merge them behind a prefix,
+        // turn them into a single node with combined number of
+        // instances. Repeat.
+        while heap.len() > 1 {
+            let left = heap.pop().unwrap();
+            let right = heap.pop().unwrap();
+            heap.push(Reverse(Node {
+                instances: left.0.instances + right.0.instances,
+                content: NodeContent::Internal(Rc::new((left.0.content, right.0.content))),
+            }));
+        }
+
+        // Convert tree into bit lengths
+        let root = heap.pop().unwrap(); // We have checked above that there is at least one value.
+        let mut bit_lengths = Vec::with_capacity(len);
+        fn aux<T>(bit_lengths: &mut Vec<(T, BitLen)>, depth: u8, node: NodeContent<T>) {
```

Comment:
Might be worth putting a FIXME/TODO and an assert here that the length generated is not >20 bits, since this is the earliest you can detect length limits are going to be violated and the code you'll have to change to do limits.

Reply:
I thought the idea was to not make the 20 bits part of the spec?

Reply:
Added a mechanism to check that we do not exceed a maximal length.